packages/xod-client/src/hinting/validators.js

```diff
@@ -0,0 +1,230 @@
+
+import {
+  generalValidator,
+  callFnIfExist,
+  mergeErrors,
+  validatePatches,
+  validateLocalPatches,
+  validateChangedPatch,
+  validatePatchesGenerally,
+  setAssocPolicy,
+  setMergePolicy,
+  // Basic validate functions:
+  getVariadicMarkersErrorMap,
+  getDeadRefErrorMap,
+  validateBoundValues,
+  validateLinkPins,
+} from './validators.internal';
+
+// PinErrors :: { errors: Map ErrorType (Maybe [Error]) }
+// LinkErrors :: { errors: Map ErrorType (Maybe [Error]) }
+// NodeErrors :: { errors: Map ErrorType (Maybe [Error]), pins: Map PinKey PinErrors }
+// PatchErrors :: { errors: Map ErrorType (Maybe [Error]), nodes: Map NodeId NodeErrors }
+
+// =============================================================================
+//
+// Predicates for preventing excessive validations
+// &
+// Short validate functions
+//
+// =============================================================================
+
+// :: Node -> Boolean
+const isNodeTerminalOrSelf = R.compose(
+  R.either(XP.isTerminalPatchPath, XP.isTerminalSelf),
+  XP.getNodeType
+);
+
+// A map of functions, that accepts new project and action
+// to check shall we need to run any validations on this changes
+// Indexed by ActionTypes
+// Return false to skip validation
+// Map ActionType (Action -> Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> Boolean)
+const shouldValidate = {
+  [PAT.BULK_MOVE_NODES_AND_COMMENTS]: (action, project) => {
+    // Could change validity only when moving terminals or `output-self` marker
+    const nodeIds = action.payload.nodeIds;
+    if (nodeIds.length === 0) return false;
+
+    return R.compose(
+      foldMaybe(false, XP.isVariadicPatch),
+      XP.getPatchByPath(action.payload.patchPath)
+    )(project);
+  },
+  [PAT.NODE_UPDATE_PROPERTY]: (action, project) => {
+    const key = action.payload.key;
+    // If User changed description — do not validate
+    if (key === 'description') return false;
+
+    const nodeId = action.payload.id;
+    const patchPath = action.payload.patchPath;
+
+    return R.compose(
+      foldMaybe(
+        true, // Very strange behaviour — let's validate
+        R.either(
+          // If changed label — do not validate
+          () => key !== 'label',
+          // But if it's terminal or self Node — validate
+          isNodeTerminalOrSelf
+        )
+      ),
+      R.chain(XP.getNodeById(nodeId)),
+      XP.getPatchByPath
+    )(patchPath, project);
+  },
+  [PAT.PATCH_DESCRIPTION_UPDATE]: R.F,
+  [PAT.PATCH_NATIVE_IMPLEMENTATION_UPDATE]: R.F,
+};
+
+/**
+ * A map of short-circuit validations and exceptions to the general rules
+ * of validation.
+ * For example of exception case see `PAT.PATCH_ADD`.
+ *
+ * Indexed by ActionTypes.
+ * Should return `ErrorsUpdateData` { policy: UPDATE_ERRORS_POLICY, errors: Map PatchPath PatchErrors }
+ *
+ * Better and simplies way to produce this type is apply `set*Policy` function
+ * as a last step of custom validation.
```

Comment:
**A** better and **simpler** way to produce this type is **to** apply `set*Policy` function as **the** last step of **a** custom validation.